contracts/remarkables/src/contract.rs

```diff
@@ -0,0 +1,699 @@
+    }
+    Ok(())
+}
+
+fn execute_update_admin(
+    deps: DepsMut<DesmosQuery>,
+    info: MessageInfo,
+    new_admin: String,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    check_admin(deps.storage, &info)?;
+    let new_admin_addr = deps.api.addr_validate(&new_admin)?;
+    CONFIG.update(deps.storage, |mut config| -> Result<_, ContractError> {
+        config.admin = new_admin_addr.clone();
+        Ok(config)
+    })?;
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_UPDATE_ADMIN)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_ADMIN, &info.sender)
+        .add_attribute(ATTRIBUTE_NEW_ADMIN, new_admin_addr))
+}
+
+fn execute_update_rarity_mint_fees(
+    deps: DepsMut<DesmosQuery>,
+    info: MessageInfo,
+    level: u32,
+    new_fees: Vec<Coin>,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    check_admin(deps.storage, &info)?;
+    RARITY.update(deps.storage, level, |rarity| -> Result<_, ContractError> {
+        let mut new_rarity = rarity.ok_or(ContractError::RarityNotExists { level })?;
+        new_rarity.mint_fees = new_fees;
+        Ok(new_rarity)
+    })?;
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_UPDATE_RARITY_MINT_FEES)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_RARITY_LEVEL, level.to_string()))
+}
+
+fn check_admin(storage: &dyn Storage, info: &MessageInfo) -> Result<(), ContractError> {
+    let config = CONFIG.load(storage)?;
+    if config.admin != info.sender {
+        return Err(ContractError::NotAdmin {
+            caller: info.sender.clone(),
+        });
+    }
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps<DesmosQuery>, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Config {} => to_binary(&query_config(deps)?),
+        QueryMsg::Rarities {} => to_binary(&query_rarities(deps)?),
+        QueryMsg::AllNftInfo {
+            token_id,
+            include_expired,
+        } => to_binary(&query_all_nft_info(deps, token_id, include_expired)?),
+        QueryMsg::Tokens {
+            owner,
+            start_after,
+            limit,
+        } => to_binary(&query_tokens(deps, owner, start_after, limit)?),
+    }
+}
+
+fn query_config(deps: Deps<DesmosQuery>) -> StdResult<QueryConfigResponse> {
+    let config = CONFIG.load(deps.storage)?;
+    let cw721_address = CW721_ADDRESS.load(deps.storage)?;
+    Ok(QueryConfigResponse {
+        admin: config.admin,
+        cw721_code_id: config.cw721_code_id.into(),
+        cw721_address,
+        subspace_id: config.subspace_id.into(),
+    })
+}
+
+fn query_rarities(deps: Deps<DesmosQuery>) -> StdResult<QueryRaritiesResponse> {
+    let res: StdResult<Vec<_>> = RARITY
+        .range(deps.storage, None, None, Order::Ascending)
+        .collect();
+    let rarities = res?
+        .iter()
```

Comment:
Use [Vec::drain](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.drain) insetad to avoid the copy of `level` and `mint_fees` inside `map`.